syntax = "proto3";
package dracon.v1;
option go_package = "v1";

import "google/protobuf/timestamp.proto";

/* Confidence represents the tool's confidence that an issue exists */
enum Confidence {
    CONFIDENCE_INFO = 0;
    CONFIDENCE_LOW = 1;
    CONFIDENCE_MEDIUM = 2;
    CONFIDENCE_HIGH = 3;
    CONFIDENCE_CRITICAL = 4;
}

/* Severity represents the reported issue severity */
enum Severity {
    SEVERITY_INFO = 0;
    SEVERITY_LOW = 1;
    SEVERITY_MEDIUM = 2;
    SEVERITY_HIGH = 3;
    SEVERITY_CRITICAL = 4;
}

/* Issue represents a vulnerability to be processed by consumers */
message Issue {
    string target = 1;      // can be host:port or //vault/foo/bar:34-67
    string type = 2;        // CWE-ID, etc for XSS, CSRF, etc.
    string title = 3;       // the vulnerability title from the tool
    Severity severity = 4;
    double cvss = 5;
    Confidence confidence = 6;
    string description = 7; // human readable description of the issue
    string source = 8; // https://github.com/thought-machine/dracon.git?ref=<revision>, github.com:tektoncd/pipeline.git?ref=<revision>, local?ref=local
    string cve = 9; // [Optional] the CVE causing this vulnerability
}

/* Represents an issue that has been enriched with metadata from the enrichment service */
message EnrichedIssue {
    Issue raw_issue = 1;
    // The first time this issue was seen by the enrichment service
    google.protobuf.Timestamp first_seen = 2;
    // The number of times this issue was seen
    uint64 count = 3;
    // Whether this issue has been previously marked as a false positive
    bool false_positive = 4;
    // The last time this issue was updated
    google.protobuf.Timestamp updated_at = 5;
    // hash
    string hash = 6;
}
